packages/keyring-controller/src/KeyringController.ts

```diff
@@ -2042,6 +2071,34 @@ export class KeyringController extends BaseController<
     this.messagingSystem.publish(`${name}:unlock`);
   }
 
+  /**
+   * Execute the given function after acquiring the controller lock
+   * and save the keyrings to state after it, or rollback to their
+   * previous state in case of error.
+   *
+   * @param fn - The function to execute.
+   * @returns The result of the function.
+   */
+  async #persistOrRollback<T>(fn: MutuallyExclusiveCallback<T>): Promise<T> {
+    return this.#withControllerLock(async ({ releaseLock }) => {
+      const currentSerializedKeyrings = await this.#getSerializedKeyrings();
+
+      try {
+        const callbackResult = await fn({ releaseLock });
+        // State is committed only if the operation is successful
+        await this.#updateVault();
+
+        return callbackResult;
+      } catch (e) {
+        // Keyrings are cleared and restored to their previous state
+        await this.#clearKeyrings({ skipStateUpdate: true });
+        await this.#restoreSerializedKeyrings(currentSerializedKeyrings);
+
+        throw e;
+      }
+    });
+  }
+
```

Comment:
Serializing and restoring (in case it's needed) should be the most reliable way to roll back all keyrings to a previous state.

Reply:
An alternative would have been something like:
```typescript
const backup = this.#keyrings.slice();
// --
this.#keyrings = backup
```
But this doesn't work for several reasons:
1. The shallow copies of keyring instances in `backup` would be different instances than the original ones (which is ok); but if a keyring instance has a class variable with an array of objects, the shallow copy of that instance would have a class variable array with references to the same objects of the original. This is the case with the HD Keyring, so a rollback would not work there
2. An eventual QRKeyring event subscription from the controller would be left invalid in case of rollback, since the controller would be subscribed to the original instance instead of the shallow copy
3. Having two living instances of a keyring could create unwanted scenarios (e.g. two keyring instances that communicate with the same device) 
4. Shallow copies could not be initialized properly